[1/2] DHCP: Refactor dhcp client code

holmanb opened the pull request:
```
Move isc-dhclient code to dhcp.py

In support of the upcoming deprecation of
isc-dhcp-client, this code refactors current
dhcp code into classes in dhcp.py. The
primary user-visible change should be the
addition of the following log:

dhcp.py[DEBUG]: DHCP client selected: dhclient

This code lays groundwork to enable
alternate implementations to live side by
side in the codebase to be selected with
distro-defined priority fallback. Note that
maybe_perform_dhcp_discovery() now selects
which dhcp client to call, and then runs the
corresponding client's dhcp_discovery()
method. Currently only class IscDhclient is
implemented, however a yet-to-be-implemented
class Dhcpcd exists to test fallback behavior
and this will be implemented in part two of
this series.

Part of this refactor includes shifting
dhclient service management from hardcoded
calls to the distro-defined manage_service()
method in the *BSDs. Future work is required
in this area to support multiple clients via
select_dhcp_client().
```

## Additional Context
Reviewers may find this pull request easiest to review commit by commit rather than all at once, since this change spans several files. Each commit is logically grouped - code changes and corresponding test updates share commits.

## Test Steps
Boot cloud-init on a datasource that uses ephemeral network setup. 

```bash
tox -e integration-test -- tests/integration_tests/test_logging.py
```

New Log:
```
2023-04-13 18:00:28,773 - util.py[DEBUG]: Read 5 bytes from /sys/class/net/ens5/operstate
2023-04-13 18:00:28,773 - dhcp.py[DEBUG]: DHCP client selected: dhclient                                                                                                                  
2023-04-13 18:00:28,773 - dhcp.py[DEBUG]: Performing a dhcp discovery on ens5  
```

[Full log](https://github.com/canonical/cloud-init/files/11225377/cloud-init.log)

## Checklist:

 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

igalic reviewed (commented):
first look

holmanb commented on cloudinit/sources/DataSourceVultr.py:
```diff
@@ -136,7 +137,16 @@ def get_datasource_list(depends):
         print("Machine is not a Vultr instance")
         sys.exit(1)
 
+    # It should probably be safe to try to detect distro via stages.Init(),
```
@eb3095 I'm not quite sure what `__main__` gets used for, or if it is even still used at all. Would you mind looking at this section if you get a chance?

Basically piping distros throughout the dhcp and ephemeral setup code means that `vultr.get_metadata()` will now require a distro object. This approach uses the same strategy that cloud-init normally uses. Will that work here? Based on the `is_vultr()` check above, I assume that this should be running on an image with cloud-init. Therefore, I assume that this approach would work, but I want to double check my assumption.

eb3095 replied:
Yeah this should work fine. __main__ i believe was an early discussion about how to debug. I dont think we have ever really used it, but that was the intended purpose.

holmanb commented on cloudinit/net/dhcp.py:
```diff
@@ -267,101 +132,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        subp.subp(["ip", "link", "set", "dev", interface, "up"], capture=True)
```
Good point. This is how it works currently on main, but you're right, this isn't going to work without iproute2. I think that means that much of `ephemeral.py` needs to be reworked too (which should finally be reasonable to do now that we're getting distros plumbed through that module).

holmanb commented on cloudinit/net/dhcp.py:
```diff
@@ -267,101 +132,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+            pid_file,
+            interface,
+            "-sf",
+            "/bin/true",
```
I'll propose something that uses the equivalents available in freebsd

holmanb commented on cloudinit/net/dhcp.py:
```diff
@@ -267,101 +132,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+
+    @staticmethod
+    def get_dhclient_d():
+        # find lease files directory
```
Good question. This is basically a direct copy/paste from DataSourceCloudStack, but modified to be in a class.

From the git history I see that:

- [6 years ago](https://github.com/canonical/cloud-init/commit/33816e96d8981918f734dab3ee1a967bce85451a) the network manager directory was added. I suspect that this is no longer required, since recently when digging around in upstream networkmanager code I recall seeing that isc-dhclient was dropped in favor of an internal implementation. A hasty grep of my local (ubuntu) NetworkManager-based system sees nothing that matches `dhcp-server-identifier`, which is the whole point of this function.
- [12 years ago](https://github.com/canonical/cloud-init/commit/4fde399a38765fa9641b3177b966ad6c8ec9750f) Ubuntu support was added, Fedora was already supported

I'm a bit hesitant to split this up, however, because of the risk of accidentally breaking one distro or another by making this more exact. Unless someone wants to audit all of the distros, I'd be more comfortable leaving it as is for now.

holmanb commented on cloudinit/net/dhcp.py:
```diff
@@ -267,101 +132,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+        #   dhclient-<iface>.leases, dhclient6.leases
+        # centos7: ('--' is not a typo)
+        #   dhclient--<iface>.lease, dhclient6.leases
+        for fname in lease_files:
```
It could be, but I'm not going to prioritize it right now. Right now my focus is getting functionality correct without risking breakage of things that currently work.

blackboxsw commented on cloudinit/net/dhcp.py:
```diff
@@ -364,4 +369,86 @@ def _trunc_error(cidr, required, remain):
+    return latest_file
+
+
+def parse_dhcp_server_from_lease_file(lease_file):
```
Is there opportunity to leverage `parse_dhcp_lease_file` here? 

We could also speed up this code potentially by reversing the order of leases processed to break on the first case of `dhcp-server-identifier` instead of processing the whole file and keeping the last case. It may not matter much in practice as the lease file processed in most cases may only have 1 lease in it though.

holmanb replied:
> Is there opportunity to leverage parse_dhcp_lease_file here?

It looks like it. This code came from a different file and looks like a case of partial code duplication.

> We could also speed up this code potentially...

Agreed. I'd like for that to be a separate PR for a cleaner git history, since it doesn't fit the scope of this change.

blackboxsw commented on cloudinit/distros/openbsd.py:
```diff
@@ -25,7 +25,7 @@ def _write_hostname(self, hostname, filename):
         return ["usermod", "-G", group_name, member_name]
 
-    def manage_service(self, action: str, service: str):
+    def manage_service(self, action: str, service: str, rcs=None):
```
Forgot the `@classmethod`

holmanb replied:
I see it in the latest version of this branch. I'm guessing this got fixed in a later commit than the one that you reviewed.

blackboxsw commented on cloudinit/net/freebsd.py:
```diff
@@ -50,10 +51,8 @@ def start_services(self, run=False):
-            subp.subp(
-                ["service", "dhclient", "stop", dhcp_interface],
-                rcs=[0, 1],
-                capture=True,
+            net.dhcp.IscDhclient.start_service(
+                dhcp_interface, distros.freebsd.Distro
             )
```
I don't think we can make this shift as-written. These definitions are no longer the same, the new `IscDhclient.start_service` method completely ignores the dhcp_interface because we are using `distro.manage_interface` and dropping the supplemental `dhcp_interface` argument in the refactor.

Instead of running:
`service dhclient stop eth0` 
we now run only the following on FreeBSD
`service dhclient stop`

holmanb replied:
Nice catch, thanks. This should be fixed in [f1eb8ac](https://github.com/canonical/cloud-init/pull/2122/commits/f1eb8acc98b0995cfb7f00c0c24adfa7f951b604).

blackboxsw commented on cloudinit/net/dhcp.py:
```diff
@@ -267,101 +131,366 @@ def networkd_get_option_from_leases(keyname, leases_d=None):
+
+    @staticmethod
+    def parse_dhcp_lease_file(lease_file):
+        """Parse the given dhcp lease file for the most recent lease.
```
This returns a list of all leases. Let's use type annotations on the function definitions here to guide our expectations about params and return types
And let's also either:
* fix the logic to return only one lease and correct the one call-site in ephemeral.py to assume most recent lease
-- or --
* Fix the docstr header

holmanb replied:
fixed in 5cd048f6311ae4c509d890d3dfb0320913b6252c

holmanb replied:
Sorry, I must have accidentally dropped that in a rebase. Thanks!

blackboxsw reviewed (commented):
Quick first pass. Wanted to get my BSD manage_service question out there as it may involve a bit of refactor to retain previous behavior. I'll continue the review today.

holmanb commented:
> Quick first pass. Wanted to get my BSD manage_service question out there as it may involve a bit of refactor to retain previous behavior. I'll continue the review today.

Thanks for the initial review @blackboxsw. I think that I have addressed all of your comments. 

Additionally I just pushed up the refactor (commit 9d16846c2b8061bf5c6bd06f413d9419ce729101) that makes network ops in ephemeral.py distro-agnostic (cc @igalic).

blackboxsw commented on cloudinit/net/dhcp.py:
```diff
@@ -167,7 +168,7 @@ def __init__(self):
 
     @staticmethod
-    def parse_dhcp_lease_file(lease_file):
+    def parse_dhcp_lease_file(lease_file: str) -> list:
```
I think we may want to `from typing import Any, Dict, List` and have to following:
```suggestion
    def parse_dhcp_lease_file(lease_file: str) -> List[Dict[str, Any]]:
```

blackboxsw commented on cloudinit/distros/__init__.py:
```diff
@@ -104,18 +111,22 @@ class Distro(persistence.CloudInitPickleMixin, metaclass=abc.ABCMeta):
     # This is used by self.shutdown_command(), and can be overridden in
     # subclasses
     shutdown_options_map = {"halt": "-H", "poweroff": "-P", "reboot": "-r"}
+    net_ops = iproute2
```
This net_ops work feels like a bolt-on ideal that represents a bit of scope creep for this PR. Would it be okay to separate this to another PR to ease review? Otherwise, we can spend a bit more time going through this in this PR but I worry a bit about missing something here in the refactor dance.

holmanb replied:
That's fair. I'll turn this into a 3 part series.

blackboxsw commented on cloudinit/net/freebsd.py:
```diff
@@ -50,10 +51,8 @@ def start_services(self, run=False):
-                ["service", "dhclient", "stop", dhcp_interface],
-                rcs=[0, 1],
-                capture=True,
+            net.dhcp.IscDhclient.start_service(
```
@holmanb it looks like this PR unintentionally inverts the operations, the intent here was to stop any existing dhcp services on any dhcp interfaces before starting them again (instead of just using the systemctl restart operation). But it looks like we are now starting and then stopping. Let's flip them back.

holmanb replied:
Nice catch. Fixed.

blackboxsw reviewed (changes requested):
Looks good @holman. Just to make sure I'm not lost in refactor intents, if we can separate net_ops work from the IscDhclient work it'd make me more certain we aren't missing something in review.

holmanb commented:
> Looks good @holman. 

Thanks! However, Zach might be confused by the Github notification from your comment ;) 

> Just to make sure I'm not lost in refactor intents, if we can separate net_ops work from the IscDhclient work it'd make me more certain we aren't missing something in review.

Nice attention to detail in the review. And I agree - breaking that out into a separate commit will make this a more coherent commit, in addition to helping review burden. I've moved that work [here](https://github.com/canonical/cloud-init/pull/2127).

blackboxsw reviewed (approved):
@holmanb LGTM, thanks for separating the PRs into something easier to review. Refactor looks good.

blackboxsw reviewed (changes requested):
hold merge on this for a moment, I want to check that Vultr invocation of `Init` I had missed.

blackboxsw reviewed (approved):
Clear, If we have to call `stages.Init()` to get the distro object for callers of `__main__`. I think that's ok. We could source this from the `sources.pkl_load(/var/lib/cloud/instance/obj.pkl)`, but I think stages.Init() is probably an easier entry point.
